### PR TITLE
TLF reidentification when needed related to rekey

### PR DESCRIPTION
**Pull request by taruti**
Rekey ends up calling getMDLocked and thus we add an identity if we didn't have the md already locally in memory.

### CODE REVIEW

**Comment by strib** [comment]
Thanks for doing this @taruti.  Is this intended to fix KBFS-1163?  I don't think it does -- it just moves the identify around to happening in `folderBranchOps` rather than `KeyManagerStandard`, but it still uses the same, possibly bogus, MD served up by the server.

What we really need is some what to verify that the `TlfHandle` associated with the MD is actually the one associated the the `TlfID`.  The parts of the MD that determine the `TlfHandle` can currently be forged by the server and so we need to somehow compare them to a handle provided by the user.  This is usually done by the fact that the user accessed the folder through a path including the set of users, but in the case of a rekey we don't have that option.

@akalin-keybase should take a look and chime in though, since he has thought about this issue more.

File: libkbfs/folder_branch_ops.go [hunk 1]
```diff
@@ -950,6 +950,10 @@ func (fbo *folderBranchOps) getMDLocked(
 	if err != nil {
 		return ImmutableRootMetadata{}, err
 	}
+	err = fbo.identifyOnce(ctx, md.ReadOnly())
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: Yeah, I don't think this is correct -- if rtype is `mdReadNoIdentify` that's an explicit signal to skip the identify.

File: libkbfs/conflict_resolver.go [hunk 2]
```diff
@@ -3770,7 +3770,7 @@ outer:
 		return err
 	}
 
-	handle := cr.fbo.getHead(lState).GetTlfHandle()
+	handle := cr.fbo.getTrustedHead(lState).GetTlfHandle()
```
Comment (akalin-keybase) [inline libkbfs/conflict_resolver.go]: seems like we should be checking here for a nil head, or somehow plumbing the name/public info through from the caller.

Comment (akalin-keybase) [inline libkbfs/conflict_resolver.go]: yeah, I think just checking for a nil head and panicking is fine

File: libkbfs/conflict_resolver.go [hunk 3]
```diff
@@ -3794,7 +3794,7 @@ func (cr *ConflictResolver) doResolve(ctx context.Context, ci conflictInput) {
 	defer func() {
 		cr.log.CDebugf(ctx, "Finished conflict resolution: %v", err)
 		if err != nil {
-			handle := cr.fbo.getHead(lState).GetTlfHandle()
+			handle := cr.fbo.getTrustedHead(lState).GetTlfHandle()
```
Comment (akalin-keybase) [inline libkbfs/conflict_resolver.go]: here, too

File: libkbfs/folder_branch_ops.go [hunk 4]
```diff
@@ -230,9 +230,10 @@ type folderBranchOps struct {
 	// should only be taken in the following order to avoid deadlock:
 	mdWriterLock leveledMutex // taken by any method making MD modifications
 
-	// protects access to head and latestMergedRevision.
-	headLock leveledRWMutex
-	head     ImmutableRootMetadata
+	// protects access to head, headTrusted and latestMergedRevision.
+	headLock    leveledRWMutex
+	head        ImmutableRootMetadata
+	headTrusted bool
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: I think trustedness is important enough to deserve its own type, say, `HeadTrustStatus` with values `HeadTrusted` and `HeadUntrusted`. It would also be a good place to put a comment about what exactly 'trusted' and 'untrusted' mean in this context.

File: libkbfs/folder_branch_ops.go [hunk 5]
```diff
@@ -493,13 +494,23 @@ func (fbo *folderBranchOps) deleteFromFavorites(ctx context.Context,
 	return favorites.Delete(ctx, h.ToFavorite())
 }
 
-// getHead should not be called outside of folder_branch_ops.go.
-func (fbo *folderBranchOps) getHead(lState *lockState) ImmutableRootMetadata {
+// getTrustedHead should not be called outside of folder_branch_ops.go.
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: mention that it returns the empty `ImmutableRootMetadata` if the head is untrusted.

File: libkbfs/folder_branch_ops.go [hunk 6]
```diff
@@ -574,6 +598,16 @@ func (fbo *folderBranchOps) setHeadLocked(
 	if !isFirstHead {
 		wasReadable = fbo.head.IsReadable()
 
+		if !fbo.headTrusted {
+			err := fbo.validateHead(ctx, md)
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: Looks like `trusted` is only ever false when setting the first head, so I think we want to make sure that `trusted` is true in the `!isFirstHead` clause.

File: libkbfs/folder_branch_ops.go [hunk 7]
```diff
@@ -656,6 +690,7 @@ func (fbo *folderBranchOps) setHeadLocked(
 	}
 
 	fbo.head = md
+	fbo.headTrusted = trusted
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: I think for clarity this should be done only if `isFirstHead` holds.

File: libkbfs/folder_branch_ops.go [hunk 8]
```diff
@@ -873,34 +909,46 @@ func (fbo *folderBranchOps) identifyOnce(
...
 }
 
-// if rtype == mdWrite || mdRekey, then mdWriterLock must be taken
-func (fbo *folderBranchOps) getMDLocked(
+// getMdLockedForRead returns an existing md for read
+// operation. Note that mds will not be fetched here.
+func (fbo *folderBranchOps) getMDLockedForRead(
 	ctx context.Context, lState *lockState, rtype mdReqType) (
 	md ImmutableRootMetadata, err error) {
+	if rtype != mdReadNeedIdentify && rtype != mdReadNoIdentify {
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: maybe split `mdReqType` into two types, one for reads and one for writes.

File: libkbfs/folder_branch_ops.go [hunk 9]
```diff
@@ -873,34 +909,46 @@ func (fbo *folderBranchOps) identifyOnce(
...
+
+	return ImmutableRootMetadata{}, MDWriteNeededInRequest{}
+}
+
+// getMDLockedForWrite can fetch MDs, identify them and
+// contains the fancy logic. For reading use getMDLockedForRead.
+// Here we actually can fetch things from the server.
+// rekeys are untrusted.
+func (fbo *folderBranchOps) getMDLockedForWriteOrRekey(
+	ctx context.Context, lState *lockState, isRekey bool) (
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: would rather have a type with `mdWrite` and `mdRekey` here.

File: libkbfs/folder_branch_ops.go [hunk 10]
```diff
@@ -873,34 +909,46 @@ func (fbo *folderBranchOps) identifyOnce(
...
-		return ImmutableRootMetadata{}, MDWriteNeededInRequest{}
-	}
-
-	// We go down this code path either due to a rekey
-	// notification for an unseen TLF, or in some tests.
-	//
-	// TODO: Make tests not take this code path, and keep track of
-	// the fact that MDs coming from rekey notifications are
-	// untrusted.
+	// We come down this path in e.g. tests.
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: is the old comment about coming down this code path due to a rekey notification for an unseen TLF not true now?

File: libkbfs/folder_branch_ops.go [hunk 11]
```diff
@@ -1018,19 +1066,33 @@ func (fbo *folderBranchOps) getMDForReadNeedIdentify(
...
 	if _, noMD := errors.Cause(err).(NoMergedMDError); noMD {
 		return ImmutableRootMetadata{}, nil
 	}
 
-	return irmd, err
+	if err != nil {
+		return ImmutableRootMetadata{}, err
+	}
+
+	if !md.TlfID().IsPublic() {
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: just curious, why did you add this check? I wonder if we should be doing this before we set the head (if we're not already).

File: libkbfs/kbfs_ops_test.go [hunk 12]
```diff
@@ -357,6 +357,7 @@ func injectNewRMD(t *testing.T, config *ConfigMock) (
 	ops := getOps(config, id)
 	ops.head = makeImmutableRMDForTest(
 		t, config, rmd, fakeMdID(tlf.FakeIDByte(id)))
+	ops.headTrusted = true
```
Comment (akalin-keybase) [inline libkbfs/kbfs_ops_test.go]: can you try writing a test that makes sure the new code behaves as expected, if an MD server sends down a rekey notification and sends down a fake TLF?

You can probably copy some of the code from `TestKBFSOpsMaliciousMDServerRange`

File: libkbfs/folder_branch_ops.go [hunk 13]
```diff
@@ -873,34 +909,46 @@ func (fbo *folderBranchOps) identifyOnce(
 	return nil
 }
 
-// if rtype == mdWrite || mdRekey, then mdWriterLock must be taken
-func (fbo *folderBranchOps) getMDLocked(
+// getMdLockedForRead returns an existing md for read
+// operation. Note that mds will not be fetched here.
+func (fbo *folderBranchOps) getMDLockedForRead(
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: the naming convention is `getMDForReadLocked`

File: libkbfs/folder_branch_ops.go [hunk 14]
```diff
@@ -873,34 +909,46 @@ func (fbo *folderBranchOps) identifyOnce(
...
+	}
+
+	return ImmutableRootMetadata{}, MDWriteNeededInRequest{}
+}
+
+// getMDLockedForWrite can fetch MDs, identify them and
+// contains the fancy logic. For reading use getMDLockedForRead.
+// Here we actually can fetch things from the server.
+// rekeys are untrusted.
+func (fbo *folderBranchOps) getMDLockedForWriteOrRekey(
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: `getMDForWriteOrRekeyLocked`

File: libkbfs/folder_branch_ops.go [hunk 15]
```diff
@@ -564,8 +575,21 @@ func (fbo *folderBranchOps) getJournalPredecessorRevision(ctx context.Context) (
 	return jStatus.RevisionStart - 1, nil
 }
 
+// validateHead should be called with headLock taken.
+func (fbo *folderBranchOps) validateHead(
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: the convention is to name this `validateHeadLocked`, have it take an `lState` variable, and call `fbo.headLock.AssertLocked(lState)`

File: libkbfs/conflict_resolver.go (hunk 2 above)
Comment (taruti) [inline libkbfs/conflict_resolver.go]: Done.

File: libkbfs/conflict_resolver.go (hunk 3 above)
Comment (taruti) [inline libkbfs/conflict_resolver.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 4 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 5 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 15 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 6 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Now panic on the logic error.

File: libkbfs/folder_branch_ops.go (hunk 7 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 13 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 14 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 8 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 9 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 10 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Resolved.

File: libkbfs/folder_branch_ops.go (hunk 11 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Previously this was going though `getMDForReadHelper` which includes the check. After separating the reads and write calls properly this is the only place where `mdWrite` was used together with `getMDForReadHelper` - so inline that check rather than creating a single use separate `getMDForWriteHelper`.

File: libkbfs/kbfs_ops_test.go (hunk 12 above)
Comment (taruti) [inline libkbfs/kbfs_ops_test.go]: Ok, will try. (not yet in the done)

File: libkbfs/folder_branch_ops.go [hunk 16]
```diff
@@ -24,18 +24,23 @@ import (
...
-	mdReadNoIdentify mdReqType = iota
+	mdReadNoIdentify mdReadType = iota
 	// A read request that needs an identify to be performed (if
 	// it hasn't been already).
 	mdReadNeedIdentify
-	// A write request.
-	mdWrite
+)
+
+// mdModificationType indicates modification type.
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: `mdUpdateType` might be a better (shorter) name

File: libkbfs/folder_branch_ops.go [hunk 17]
```diff
@@ -165,6 +170,17 @@ func (bl *blockLock) DoRUnlockedIfPossible(lState *lockState, f func(*lockState)
 	f(lState)
 }
 
+// headTrustStatus marks whether the head is from a trusted or
+// untrusted source. When rekeying we get the fbo by folder
+// id and do not check the tlf id, thus resulting in untrusted
+// fbo heads.
+type headTrustStatus bool
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: i don't think this is desirable, since then `headTrustStatus` can be used in an `if` statement without comparing it to a constant value. int would be better.

File: libkbfs/folder_branch_ops.go [hunk 18]
```diff
@@ -165,6 +170,17 @@ func (bl *blockLock) DoRUnlockedIfPossible(lState *lockState, f func(*lockState)
...
+
+const (
+	headTrusted   = headTrustStatus(true)
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: `headTrusted headTrustStatus = ...`

File: libkbfs/folder_branch_ops.go [hunk 19]
```diff
@@ -230,9 +246,10 @@ type folderBranchOps struct {
 	// should only be taken in the following order to avoid deadlock:
 	mdWriterLock leveledMutex // taken by any method making MD modifications
 
-	// protects access to head and latestMergedRevision.
-	headLock leveledRWMutex
-	head     ImmutableRootMetadata
+	// protects access to head, headTrusted and latestMergedRevision.
+	headLock    leveledRWMutex
+	head        ImmutableRootMetadata
+	headTrusted headTrustStatus
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: maybe `headStatus` as a better name (since `headTrusted` still sounds boolean-ish)

File: libkbfs/folder_branch_ops.go [hunk 20]
```diff
@@ -564,16 +594,48 @@ func (fbo *folderBranchOps) getJournalPredecessorRevision(ctx context.Context) (
 	return jStatus.RevisionStart - 1, nil
 }
 
-func (fbo *folderBranchOps) setHeadLocked(
+// validateHeadLocked validates an untrusted head and sets it as trusted.
+// see headTrustedState comment for more information.
+// validateHeadLocked should be called with headLock taken.
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: can remove this last line, since the assertion below mentions it

File: libkbfs/folder_branch_ops.go [hunk 21]
```diff
@@ -564,16 +594,48 @@ func (fbo *folderBranchOps) getJournalPredecessorRevision(ctx context.Context) (
...
+func (fbo *folderBranchOps) validateHeadLocked(
 	ctx context.Context, lState *lockState, md ImmutableRootMetadata) error {
+
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: remove extra line?

File: libkbfs/folder_branch_ops.go [hunk 22]
```diff
@@ -564,16 +594,48 @@ func (fbo *folderBranchOps) getJournalPredecessorRevision(ctx context.Context) (
...
+		fbo.log.CCriticalf(ctx, "Fake untrusted TLF encountered %v %v %v %v", fbo.head.TlfID(), md.TlfID(), fbo.head.mdID, md.mdID)
+		return MDTlfIDMismatch{fbo.head.TlfID(), md.TlfID()}
+	}
+	fbo.headTrusted = headTrusted
+	return nil
+}
+
+func (fbo *folderBranchOps) setHeadLocked(
+	ctx context.Context, lState *lockState,
+	md ImmutableRootMetadata, trusted bool) error {
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: take `headTrustStatus`?

File: libkbfs/folder_branch_ops.go [hunk 23]
```diff
@@ -1356,8 +1449,8 @@ func (fbo *folderBranchOps) SetInitialHeadFromServer(
 	// (e.g., calling `identifyOnce` and downloading the merged
 	// head) if head is already set.
 	lState := makeFBOLockState()
-	head := fbo.getHead(lState)
-	if head != (ImmutableRootMetadata{}) && head.mdID == md.mdID {
+	head, trusted := fbo.getHead(lState)
+	if trusted && head != (ImmutableRootMetadata{}) && head.mdID == md.mdID {
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: compare explicitly to a `headTrustStatus` constant (will have to do once you change the underlying type to int)

File: libkbfs/kbfs_ops_test.go (hunk 12 above)
Comment (akalin-keybase) [inline libkbfs/kbfs_ops_test.go]: If it's too hard, just add a TODO somewhere.

File: libkbfs/folder_branch_ops.go (hunk 16 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 17 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 18 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done

File: libkbfs/folder_branch_ops.go (hunk 19 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 20 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 21 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 22 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 23 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

**Comment by taruti** [comment]
Rebased and fixed things.

File: libkbfs/folder_branch_ops.go (hunk 16 above)
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: Fix comment, too.

File: libkbfs/folder_branch_ops.go [hunk 24]
```diff
@@ -230,9 +246,11 @@ type folderBranchOps struct {
 	// should only be taken in the following order to avoid deadlock:
 	mdWriterLock leveledMutex // taken by any method making MD modifications
 
-	// protects access to head, latestMergedRevision and hasBeenCleared.
-	headLock leveledRWMutex
-	head     ImmutableRootMetadata
+	// protects access to head, latestMergedRevision,
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: nit: please list in declaration order

File: libkbfs/folder_branch_ops.go [hunk 25]
```diff
@@ -496,13 +514,26 @@ func (fbo *folderBranchOps) deleteFromFavorites(ctx context.Context,
 	return favorites.Delete(ctx, h.ToFavorite())
 }
 
-// getHead should not be called outside of folder_branch_ops.go.
-func (fbo *folderBranchOps) getHead(lState *lockState) ImmutableRootMetadata {
+// getTrustedHead should not be called outside of folder_branch_ops.go.
+// Returns nil when the head is not trusted. See the comment on headTrustedStatus
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: nil -> ImmutableRootMetadata{}

File: libkbfs/folder_branch_ops.go [hunk 26]
```diff
@@ -567,16 +598,46 @@ func (fbo *folderBranchOps) getJournalPredecessorRevision(ctx context.Context) (
...
+		fbo.log.CCriticalf(ctx, "Fake untrusted TLF encountered %v %v %v %v", fbo.head.TlfID(), md.TlfID(), fbo.head.mdID, md.mdID)
+		return MDTlfIDMismatch{fbo.head.TlfID(), md.TlfID()}
+	}
+	fbo.headStatus = headTrusted
+	return nil
+}
+
+func (fbo *folderBranchOps) setHeadLocked(
+	ctx context.Context, lState *lockState,
+	md ImmutableRootMetadata, trusted headTrustStatus) error {
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: trusted -> headStatus

File: libkbfs/folder_branch_ops.go [hunk 27]
```diff
@@ -1360,8 +1456,8 @@ func (fbo *folderBranchOps) SetInitialHeadFromServer(
 	// (e.g., calling `identifyOnce` and downloading the merged
 	// head) if head is already set.
 	lState := makeFBOLockState()
-	head := fbo.getHead(lState)
-	if head != (ImmutableRootMetadata{}) && head.mdID == md.mdID {
+	head, trusted := fbo.getHead(lState)
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: trusted -> headStatus (everywhere else, too)

**Comment by akalin-keybase** [comment]
@strib maybe you should take a quick look in case i missed something

File: libkbfs/folder_branch_ops.go (hunk 16 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 24 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 25 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

File: libkbfs/folder_branch_ops.go (hunk 26 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done in all places.

File: libkbfs/folder_branch_ops.go (hunk 27 above)
Comment (taruti) [inline libkbfs/folder_branch_ops.go]: Done.

**Comment by taruti** [comment]
Broken by changes in master, investigating.

File: libkbfs/folder_branch_ops.go [hunk 28]
```diff
@@ -876,35 +941,48 @@ func (fbo *folderBranchOps) identifyOnce(
 	return nil
 }
 
-// if rtype == mdWrite || mdRekey, then mdWriterLock must be taken
-func (fbo *folderBranchOps) getMDLocked(
-	ctx context.Context, lState *lockState, rtype mdReqType) (
+// getMDForReadLocked returns an existing md for read
```
Comment (strib) [inline libkbfs/folder_branch_ops.go]: "for read" -> "for a read"

File: libkbfs/folder_branch_ops.go [hunk 29]
```diff
@@ -947,7 +1025,11 @@ func (fbo *folderBranchOps) getMDLocked(
 
 	fbo.headLock.Lock(lState)
 	defer fbo.headLock.Unlock(lState)
-	err = fbo.setInitialHeadUntrustedLocked(ctx, lState, md)
+	var headStatus = headTrusted
```
Comment (strib) [inline libkbfs/folder_branch_ops.go]: Why use `var` instead of `:=`?

File: libkbfs/folder_branch_ops.go [hunk 30]
```diff
@@ -5593,6 +5679,7 @@ func (fbo *folderBranchOps) ClearPrivateFolderMD(ctx context.Context) {
 	}
 
 	fbo.head = ImmutableRootMetadata{}
+	fbo.headStatus = headTrusted
```
Comment (strib) [inline libkbfs/folder_branch_ops.go]: Wouldn't `headUntrusted` make more sense here?  We basically want to set it back to the default value.
